usersimcrs/simulator/neural/tus/tus_feature_handler.py

```diff
@@ -0,0 +1,516 @@
+
+        Args:
+            agent_dialogue_acts: Agent dialogue acts.
+
+        Returns:
+            Feature vector for agent action.
+        """
+        v_agent_action = {intent: [0] * 3 for intent in self._agent_actions}
+        for dialogue_act in agent_dialogue_acts:
+            intent_label = dialogue_act.intent.label
+            if intent_label in self._agent_actions:
+                if not dialogue_act.annotations:
+                    v_agent_action[intent_label][0] = 1
+                for annotation in dialogue_act.annotations:
+                    if annotation.slot and annotation.value:
+                        v_agent_action[intent_label][2] = 1
+                    elif annotation.slot and annotation.value is None:
+                        v_agent_action[intent_label][1] = 1
+        return sum(v_agent_action.values(), [])
+
+    def get_slot_index_feature(self, slot: str) -> FeatureVector:
+        """Builds feature vector for slot index.
+
+        Args:
+            slot: Slot.
+
+        Returns:
+            Feature vector for slot index.
+        """
+        v_slot_index = [0] * len(self.slot_index)
+        v_slot_index[self.slot_index[slot]] = 1
+        return v_slot_index
+
+    def get_slot_feature_vector(
+        self,
+        slot: str,
+        previous_state: DialogueState,
+        state: DialogueState,
+        information_need: InformationNeed,
+        agent_dialogue_acts: List[DialogueAct],
+        user_action_vector: torch.Tensor = None,
+    ) -> FeatureVector:
+        """Builds the feature vector for a slot.
+
+        It concatenate the basic information, user action, agent action, and
+        slot index feature vectors.
+
+        Args:
+            slot: Slot.
+            previous_state: Previous state.
+            state: Current state.
+            information_need: Information need.
+            agent_dialogue_acts: Agent dialogue acts.
+            user_action_vector: User action feature vector (output vector for
+              previous turn). Defaults to None.
+
+        Returns:
+            Feature vector for the slot.
+        """
+        v_user_action = (
+            user_action_vector
+            if user_action_vector is not None
+            else torch.tensor([0] * 6)
+        )
+        _agent_dialogue_acts = []
+        # Filter agent dialogue acts by slot
+        for dialogue_act in agent_dialogue_acts:
+            for annotation in dialogue_act.annotations:
+                if annotation.slot == slot or annotation.slot is None:
+                    _agent_dialogue_acts.append(dialogue_act)
+
+        return (
+            [0, 0]  # No special token
+            + self.get_basic_information_feature(
+                slot, information_need, state, previous_state
+            )
+            + self.get_agent_action_feature(_agent_dialogue_acts)
+            + v_user_action.tolist()
+            + self.get_slot_index_feature(slot)
+        )
+
+    def get_feature_vector(
```

Comment:
Since it's multiple feature vectors, consider renaming to `get_turn_feature_vectors()`

Reply:
Done.